apm-agent-plugins/apm-java-concurrent-plugin/src/main/java/co/elastic/apm/agent/concurrent/JavaConcurrent.java

```diff
@@ -0,0 +1,229 @@
+
+import co.elastic.apm.agent.bci.ElasticApmAgent;
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.TraceContext;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ForkJoinTask;
+
+public class JavaConcurrent {
+
+    private static final WeakConcurrentMap<Object, AbstractSpan<?>> contextMap = new WeakConcurrentMap<Object, AbstractSpan<?>>(false);
+    private static final List<Class<? extends ElasticApmInstrumentation>> RUNNABLE_CALLABLE_FJTASK_INSTRUMENTATION = Collections.
+        <Class<? extends ElasticApmInstrumentation>>singletonList(RunnableCallableForkJoinTaskInstrumentation.class);
+    private static final ThreadLocal<Boolean> needsContext = new ThreadLocal<>();
+
+    private static void removeContext(Object o) {
+        AbstractSpan<?> context = contextMap.remove(o);
+        if (context != null) {
+            context.decrementReferences();
+        }
+    }
+
+    @Nullable
+    public static AbstractSpan<?> restoreContext(Object o, @Nullable ElasticApmTracer tracer) {
+        if (tracer == null) {
+            return null;
+        }
+        // When an Executor executes directly on the current thread we need to enable this thread for context propagation again
+        needsContext.set(Boolean.TRUE);
+        AbstractSpan<?> context = contextMap.remove(o);
+        if (context == null) {
+            return null;
+        }
+        if (tracer.getActive() != context) {
+            context.activate();
+            context.decrementReferences();
+            return context;
+        } else {
+            context.decrementReferences();
+            return null;
+        }
+    }
+
+    /**
+     * Wraps the provided runnable and makes this {@link AbstractSpan} active in the {@link Runnable#run()} method.
+     *
+     * <p>
+     * Note: does activates the {@link AbstractSpan} and not only the {@link TraceContext}.
+     * This should only be used when the span is closed in thread the provided {@link Runnable} is executed in.
+     * </p>
+     */
+    @Nullable
+    public static Runnable withContext(@Nullable Runnable runnable, @Nullable ElasticApmTracer tracer) {
+        if (runnable instanceof RunnableLambdaWrapper || runnable == null || tracer == null || needsContext.get() == Boolean.FALSE) {
+            return runnable;
+        }
+        needsContext.set(Boolean.FALSE);
+        AbstractSpan<?> active = tracer.getActive();
+        if (active == null) {
+            return runnable;
+        }
+        if (isLambda(runnable)) {
+            runnable = new RunnableLambdaWrapper(runnable);
```

Comment:
What about pooling wrappers? Could be used very frequently.

Reply:
By putting the runnable in a `WeakConcurrentMap`, we're already allocating a `WeakReference` and a `Map.Entry` so pooling the Runnable wrapper won't make it allocation free, unfortunately. Given that, I don't think it's worth it.